### ISSUE TITLE
[KEP-4412] Enable dynamic configuration of service account names and audiences for token requests in node audience restriction

### ISSUE THREAD

**Issue by aramase** [kubernetes/kubernetes#128678]
https://github.com/kubernetes/kubernetes/pull/128077 has been merged, and starting in v1.32, we are enforcing restrictions on the audiences for which the kubelet can request tokens.

Add new KAS flag that allows configuring list of permitted audiences specifically for image pulls.

/sig auth
/kind feature
/assign

**Comment by liggitt** [comment]
For visibility, https://github.com/kubernetes/kubernetes/pull/128077 broke one of our node daemons using kubelet credentials to provision workload identity tokens for a workload identity audience. Before the escape hatch graduates / locks on, we either need a reasonable path to giving scoped service account token creation access to a daemonset (something along the lines of https://github.com/kubernetes/enhancements/issues/4935) or a way to indicate audiences that should be allowed in addition to the API-visible ones (maybe a more general form of what this issue is tracking).

**Comment by ahmedtd** [comment]
I think a kube-apiserver flag would be sufficient for our purposes in GKE, but in general, users may run arbitrary daemonsets that reuse kubelet credentials to take advantage of node restriction.  These daemonsets might be minting tokens to talk to other external services (and thus need to set their own audiences).

So we probably need either:

* A mechanism accessible to the "cluster administrator" persona (as distinct from the "platform provider" persona) to set a list of allowed audiences for node identities to request, or
* Some sort of first-class node-restricted daemonset feature so that they no longer have to steal Kubelet's identity.

**Comment by dims** [comment]
cc @micahhausler

**Comment by mtaufen** [comment]
(I think we still need the kube-apiserver flag, but maybe ServiceAccount can be the place for custom audience config)

**Comment by enj** [comment]
> broke one of our node daemons using kubelet credentials to provision workload identity tokens for a workload identity audience

Could the associated pod be updated to include a projected SA token volume with the desired audience?

**Comment by mtaufen** [comment]
That might work in common cases, but I think there may be some edge cases with our daemon where the audiences are dynamic, depending on annotations on other resources or some other config, where we wouldn't want to redeploy the daemon just to allow a new audience. @ahmedtd (I admit, restarting kube-apiserver isn't great either).

An extension of ServiceAccount might work for these cases as well...

**Comment by enj** [comment]
@mtaufen @ahmedtd regarding the case where you have a daemon using node creds to provision SA tokens, is it safe to assume that the daemon is getting those SA tokens for it own SA (i.e. only the audience is dynamic and not the SA itself)?  If so, couldn't you simply:

1. Create a role / role binding to grant that daemon's SA the ability to create tokens for itself (these would have no audience restriction because the caller isn't a node)
2. Let the kubelet provision the daemon's SA token as normal
3. Use the daemon's SA token to create more tokens for the daemon SA with any audience you need

I suspect this would be a minor change to the daemon's code (it would have two REST clients instead of one) but it would work on any version of Kube.

**Comment by ahmedtd** [comment]
No, the SA is dynamic.  The daemon in question is serving a local RPC interface that other pods on the node use to retrieve their *own* credentials.

So if a pod running as service account `foo/bar` makes a request to the daemon, then the daemon requests a token for `foo/bar` with a specific, fixed audience.

(The token is then exchange for a Google access token representing the pod's identity in GCP IAM).

**Comment by enj** [comment]
Okay, thanks for the info, I understand the flow now.  @aramase and I have worked out a lightweight approach to allow for both dynamic service accounts and dynamic audiences.  Will open PRs after the holiday season.

**Comment by aramase** [comment]
We have a PR open for this now: https://github.com/kubernetes/kubernetes/pull/130485.
